credential/util/util.go

```diff
@@ -7,59 +7,93 @@ import (
 	"github.com/TBD54566975/ssi-sdk/credential/signing"
 	"github.com/goccy/go-json"
+	"github.com/lestrrat-go/jwx/v2/jwt"
 	"github.com/pkg/errors"
 )
 
-// CredentialsFromInterface turn a generic cred into a known shape without maintaining the proof/signature wrapper
-func CredentialsFromInterface(genericCred any) (*credential.VerifiableCredential, error) {
+// ToCredential turn a generic cred into its known object model
+func ToCredential(genericCred any) (*credential.VerifiableCredential, error) {
 	switch genericCred.(type) {
+	case *credential.VerifiableCredential:
+		return genericCred.(*credential.VerifiableCredential), nil
+	case credential.VerifiableCredential:
+		verifiableCredential := genericCred.(credential.VerifiableCredential)
+		return &verifiableCredential, nil
 	case string:
 		// JWT
-		_, _, cred, err := signing.ParseVerifiableCredentialFromJWT(genericCred.(string))
+		_, _, parsedCred, err := signing.ParseVerifiableCredentialFromJWT(genericCred.(string))
 		if err != nil {
-			return nil, errors.Wrap(err, "could not parse credential from JWT")
+			return nil, errors.Wrap(err, "parsing credential from JWT")
 		}
-		return cred, nil
+		return parsedCred, nil
 	case map[string]any:
-		// JSON
-		var cred credential.VerifiableCredential
-		credMapBytes, err := json.Marshal(genericCred.(map[string]any))
-		if err != nil {
-			return nil, errors.Wrap(err, "could not marshal credential map")
+		// VC or JWTVC JSON
+		credJSON := genericCred.(map[string]any)
+		credMapBytes, marshalErr := json.Marshal(credJSON)
+		if marshalErr != nil {
+			return nil, errors.Wrap(marshalErr, "marshalling credential map")
 		}
-		if err = json.Unmarshal(credMapBytes, &cred); err != nil {
-			return nil, errors.Wrap(err, "could not unmarshal credential map")
+
+		// first try as a VC object
+		var cred credential.VerifiableCredential
+		if err := json.Unmarshal(credMapBytes, &cred); err != nil || cred.IsEmpty() {
+			// if that fails, try as a JWT
+			_, vcFromJWT, err := VCJWTJSONToVC(credMapBytes)
+			if err != nil {
+				return nil, errors.Wrap(err, "parsing generic credential as either VC or JWT")
+			}
+			return vcFromJWT, nil
 		}
 		return &cred, nil
-	case credential.VerifiableCredential:
-		// VerifiableCredential
-		cred := genericCred.(credential.VerifiableCredential)
-		return &cred, nil
-	default:
-		return nil, fmt.Errorf("invalid credential type: %s", reflect.TypeOf(genericCred).Kind().String())
 	}
+	return nil, fmt.Errorf("invalid credential type: %s", reflect.TypeOf(genericCred).Kind().String())
 }
 
-// ClaimAsJSON converts a claim with an unknown any into the go-json representation of that credential.
-// claim can only be of type {string, map[string]interface, VerifiableCredential}.
-func ClaimAsJSON(claim any) (map[string]any, error) {
-	switch c := claim.(type) {
+// ToCredentialJSONMap turn a generic cred into a JSON object
+func ToCredentialJSONMap(genericCred any) (map[string]any, error) {
```

Comment:
Alternativly we could make this private and then the master ToCredential funciton will have the signature:

func ToCredential(genericCred any) (*credential.VerifiableCredential, map[string]any, jwt.Token, error) { 

One func to rule them all

Reply:
I had it like this at first but figured it was cleaner to have each func do a single thing